File: Behavioral/Lab22/C#/TemplateMethods/AbstractTemplate.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace C_.TemplateMethods
{
    public abstract class AbstractTemplate
    {
        protected string[] _params;
        public AbstractTemplate(params string[] args){
            _params = args;
        }
        public string[] Params{get; set;}

        public string Execute(){
            if(_params.Length == 0)
                return AddEmpty();
            else 
                return AddEstate();
        }

        protected abstract string AddEmpty();
        protected abstract string AddEstate();

        protected virtual string ReturnError() {
            return "Failed to add new Estate with args: " + string.Join(", ", _params);
        }
    }
}

File: Behavioral/Lab22/Typescript/src/TemplateMethods/abstract_template.ts
import { ICommand } from "../Interfaces/command";

export abstract class AbstractTemplate {
    protected _params: string[];
    constructor(...args: string[]) {
        this._params = args;
    }

    public get Params(): string[] {
        return this._params;
    }

    public set Params(value: string[]) {
        this._params = value;
    }

    public execute(): string{
        if(this.Params.length === 0)
            return this.addEmpty();
        else
            return this.addEstate();
    }

    protected abstract addEmpty(): string;
    protected abstract addEstate(): string;

    protected returnError(): string {
        return "Failed to add new Estate with args: " + this.Params.join(",");
    }   
}

